libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -97,321 +97,412 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
+    _attitude_target_quat.from_rotation_matrix(_ahrs.get_rotation_body_to_ned());
+    _attitude_target_quat.to_euler(_attitude_target_euler_angle.x, _attitude_target_euler_angle.y, _attitude_target_euler_angle.z);
+    _attitude_target_ang_vel = _ahrs.get_gyro();
+    _attitude_target_euler_angle = Vector3f(_ahrs.roll, _ahrs.pitch, _ahrs.yaw);
+
     // Set reference angular velocity used in angular velocity controller equal
     // to the input angular velocity and reset the angular velocity integrators.
     // This zeros the output of the angular velocity controller.
-    _ang_vel_target_rads = _ahrs.get_gyro();
+    _rate_target_ang_vel = _ahrs.get_gyro();
     get_rate_roll_pid().reset_I();
     get_rate_pitch_pid().reset_I();
     get_rate_yaw_pid().reset_I();
-
-    // Write euler derivatives derived from vehicle angular velocity to
-    // _att_target_euler_rate_rads. This resets the state of the input shapers.
-    ang_vel_to_euler_rate(Vector3f(_ahrs.roll,_ahrs.pitch,_ahrs.yaw), _ang_vel_target_rads, _att_target_euler_rate_rads);
 }
 
-void AC_AttitudeControl::shift_ef_yaw_target(float yaw_shift_cd)
+// Command a Quaternion attitude with feedforward and smoothing
+void AC_AttitudeControl::input_quaternion(Quaternion attitude_desired_quat, float smoothing_gain)
 {
-    _att_target_euler_rad.z = wrap_2PI(_att_target_euler_rad.z + radians(yaw_shift_cd*0.01f));
-}
+    // calculate the attitude target euler angles
+    _attitude_target_quat.to_euler(_attitude_target_euler_angle.x, _attitude_target_euler_angle.y, _attitude_target_euler_angle.z);
 
-void AC_AttitudeControl::input_euler_angle_roll_pitch_euler_rate_yaw_smooth(float euler_roll_angle_cd, float euler_pitch_angle_cd, float euler_yaw_rate_cds, float smoothing_gain)
-{
-    // Convert from centidegrees on public interface to radians
-    float euler_roll_angle_rad = radians(euler_roll_angle_cd*0.01f);
-    float euler_pitch_angle_rad = radians(euler_pitch_angle_cd*0.01f);
-    float euler_yaw_rate_rads = radians(euler_yaw_rate_cds*0.01f);
+    // ensure smoothing gain can not cause overshoot
+    smoothing_gain = constrain_float(smoothing_gain,1.0f,1/_dt);
 
-    // Sanity check smoothing gain
-    smoothing_gain = constrain_float(smoothing_gain,1.0f,50.0f);
+    Quaternion attitude_error_quat = _attitude_target_quat.inverse() * attitude_desired_quat;
+    Vector3f attitude_error_angle;
+    attitude_error_quat.to_axis_angle(attitude_error_angle);
 
-    // Add roll trim to compensate tail rotor thrust in heli (will return zero on multirotors)
-    euler_roll_angle_rad += get_roll_trim_rad();
-
-    if ((get_accel_roll_max_radss() > 0.0f) && _rate_bf_ff_enabled) {
-        // When roll acceleration limiting and feedforward are enabled, the sqrt controller is used to compute an euler roll-axis
-        // angular velocity that will cause the euler roll angle to smoothly stop at the input angle with limited deceleration
+    if (_rate_bf_ff_enabled & !_use_input_shaping) {
```

Comment:
Is this correct? When **not** using input shaping we enter this if? Seems weird since we then call _input_shaping_angle_...